Add support for keyboard events

expez opened the pull request:
My intention was that any keypress other than 'q' would be a NoOp, but at present all keypresses seem to either activiate the QuitCommand or cause the program to exit by some other manner.  I don't think that minor bug should hold up this PR however, in case someone else wants to work on Hatter.

expez committed:
Implement CommandFactory.

We create new commands to execute using the factory pattern.

expez committed:
Implement the NoOp command.

The Commands are in the style of the Command pattern.  Each command has
an execute method and, if applicable, an unexecute method which is used
to reverse the effect of the command.

By using the command pattern in this manner we can keep track of the
last X commands and allow the user to undo his actions, increasing
usability.

expez committed:
Add support for key events.

expez committed:
Implement QuitCommand.

expez committed:
Add documentation to Terminal and fix indentation.

jfredett commented on lib/command_factory.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  # Instantiates the command object corresponding to `key`.
+
```
It's okay to abut these lines, at the very least, they should be connected with a `#` on the empty line. I point it out because rdoc can be a finicky tool sometimes, and it might not catch the actual documentation bits.

jfredett commented on lib/commands/command.rb:
```diff
@@ -0,0 +1,9 @@
+
+  def unexcecute
+  end
+
```
You might consider not using a parent class here, and instead using a module, or just using the same interface for your commands -- premature abstraction being the root of all meta-evil, I'd opt for the latter as more ruby-ish.

jfredett commented on lib/terminal.rb:
```diff
@@ -36,6 +46,10 @@ def initialize_views
   end
 
+  def ascii_to_symbol(i)
+    i.chr.intern
```
Careful with calls to `#intern` -- it might not be a problem here, but symbols (which is what `#intern` gives you) are not GC'd, so if you end up with lots of them, it can cause memory leaks. Here, you should always be getting an integer, so we're probably fine.

I'd also use `#to_sym` instead of `#intern` -- nobody really uses the latter.

jfredett commented on spec/command_factory_spec.rb:
```diff
@@ -0,0 +1,15 @@
+  subject { CommandFactory.new nil}
+
+  it "Creates instances of type Command" do
+    subject.command(:foo).should be_a(Command)
```
This should use the short-syntax version, particularly because in newer versions of rspec, direct references to the subject are a no-go.
eg:

```
subject(:factory) { CommandFactory.new nil }
describe 'building a command' do
  subject { factory.command(:foo) }
  it { should be_a Command }
end

describe "building a command from a key" do
  let(:key) { :q }
  subject { factory.command(key) }

  it { should be_a QuitCommand }
end
```

Also be wary of `be_a` in tests, you're better off testing what methods/api it supports than what class it is. This makes your tests less brittle wrt to renaming and also opens up a lot of possibilities to share testing code in the form of shared examples, a la

```
shared_examples_for 'a command' do
  it { should respond_to :execute }
  it { should respond_to :unexecute }
  # etc
end

# ... 

describe QuitCommand do
  it_should_behave_like "a command"
  # other tests relating to QuitCommand-specific behavior
end
```

jfredett commented on spec/commands/quit_spec.rb:
```diff
@@ -0,0 +1,13 @@
+  it "sends :shutdown to the Terminal" do
+    terminal = double("terminal")
+    terminal.stub(:shutdown) { true }
+    terminal.should_receive(:shutdown)
```
rspec-spies is a nice gem for this sort of mocking, makes the setup/teardown a bit nicer. Also look into using `lets` and `befores` to do some of the lifting, w/ rspec-spies, it'd look like:

```
subject(:terminal) { double('terminal', shutdown: true) }

let(:command) { QuitCommand.new terminal }
before { command.execute }

it { should have_received :shutdown }
```

jfredett commented:
I gave it a once-over, githubs markup is fucked up for me right now so everything is in about 30pt font. I'll take a look at your comments more directly when I figure out what's happened.

expez committed:
Refactor based on product owner's suggestions:

- Remove command class, which served only the purpose of interface
  declaration.
- Use rspec-spies to create better mocks
- Create a shared example group for commands.
- Use `let' to set up context instead of creating temporary variables
  inside example groups.
- Wrap the terminal stuff so it can be elegantly skipped during testing.
- Fix formatting mistake where an empty line was inserted in an rdoc
  comment.
- Use to_sym and not intern; this is not lisp.

expez commented:
``` ruby

describe CommandFactory.new do
  ...

  ...

  describe "building a command from a key" do
    let(:key) { :q }
    subject { factory.command(key) }
    it { should be_a QuitCommand }
  end
end
```

This kind of thing will run just fine with `rspec -f progress`, but if you run it with `rspec -f documentation` it will look like 

``` sh
CommandFactory
 does this
 and that

 building a command from key
   should be a QuitCommand
```

I hav solved this by putting in an extra `context "the returned command"`.  But when reading the code it is pretty clear that the subject has changed--it happened on the line above the assertion!--so adding the extra context is really just for people running the specs with respec -f doc.  Is it common to just skip the extra embellishment and not worry about the rspec documentation output not reading like the great american novel?

jfredett commented on spec/command_factory_spec.rb:
```diff
@@ -0,0 +1,21 @@
+  let(:key) { :q }
+
+  context "Instantiates the commands matching the user's input." do
+    context "The returned command" do
```
Doubling up the context like this _is_ a bit weird, but I get the feeling you're word-smithing for description output. I'd recommend using a `describe` instead of `context` here in that case, it reads a bit better. Not a blocker.

jfredett commented on spec/commands/quit_spec.rb:
```diff
@@ -0,0 +1,20 @@
+
+  context "After executing" do
+    context "the terminal" do
+      before { command.execute }
```
Ditto above -- s/context/describe -- not a blocker

jfredett commented:
Looks generally good, I want to give it a run before I merge it, so it'll probably wait till tomorrow.

expez committed:
Refactor specs.

- Don't nest `context` blocks.  Instead, combine `context` and `describe`
to increase readabiliy.
- Be consistent about use of capitalization in specs: only top level
forms are capitalized.
